// @polkadot/types-codec/base/Base.js
import _classPrivateFieldLooseBase from "@babel/runtime/helpers/esm/classPrivateFieldLooseBase";
import _classPrivateFieldLooseKey from "@babel/runtime/helpers/esm/classPrivateFieldLooseKey";

var _registry = /*#__PURE__*/_classPrivateFieldLooseKey("registry");

var _raw = /*#__PURE__*/_classPrivateFieldLooseKey("raw");

// Copyright 2017-2022 @polkadot/types-codec authors & contributors
// SPDX-License-Identifier: Apache-2.0

/**
 * @name Base
 * @description A type extends the Base class, when it holds a value
 */
export class Base {
  constructor(registry, value, initialU8aLength) {
    Object.defineProperty(this, _registry, {
      writable: true,
      value: void 0
    });
    this.createdAtHash = void 0;
    this.initialU8aLength = void 0;
    Object.defineProperty(this, _raw, {
      writable: true,
      value: void 0
    });
    _classPrivateFieldLooseBase(this, _registry)[_registry] = registry;
    _classPrivateFieldLooseBase(this, _raw)[_raw] = value;
    this.initialU8aLength = initialU8aLength;
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    return this.toU8a().length;
  }
  /**
   * @description returns a hash of the contents
   */


  get hash() {
    return this.registry.hash(this.toU8a());
  }

  get inner() {
    return _classPrivateFieldLooseBase(this, _raw)[_raw];
  }
  /**
   * @description Checks if the value is an empty value
   */


  get isEmpty() {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].isEmpty;
  }

  get registry() {
    return _classPrivateFieldLooseBase(this, _registry)[_registry];
  }
  /**
   * @description Compares the value of the input to see if there is a match
   */


  eq(other) {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].eq(other);
  }
  /**
   * @description Returns a hex string representation of the value. isLe returns a LE (number-only) representation
   */


  toHex(isLe) {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].toHex(isLe);
  }
  /**
   * @description Converts the Object to to a human-friendly JSON, with additional fields, expansion and formatting of information
   */


  toHuman(isExtended) {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].toHuman(isExtended);
  }
  /**
   * @description Converts the Object to JSON, typically used for RPC transfers
   */


  toJSON() {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].toJSON();
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].toString();
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   * @param isBare true when the value has none of the type-specific prefixes (internal)
   */


  toU8a(isBare) {
    return _classPrivateFieldLooseBase(this, _raw)[_raw].toU8a(isBare);
  }
  /**
   * @description Returns the base runtime type name for this instance
   */


  toRawType() {
    return 'Base';
  }

  unwrap() {
    return _classPrivateFieldLooseBase(this, _raw)[_raw];
  }

}
